sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataOutboundAggregator.java

```diff
@@ -0,0 +1,357 @@
+
+  public static final String DATA_BUFFER_SIZE_LIMIT = "data_buffer_size_limit=";
+  public static final int DEFAULT_BUFFER_LIMIT_BYTES = 1_000_000;
+  public static final String DATA_BUFFER_TIME_LIMIT_MS = "data_buffer_time_limit_ms=";
+  public static final long DEFAULT_BUFFER_LIMIT_TIME_MS = -1L;
+
+  private static final Logger LOG = LoggerFactory.getLogger(BeamFnDataOutboundAggregator.class);
+  private final int sizeLimit;
+  private final long timeLimit;
+  private final Supplier<String> processBundleRequestIdSupplier;
+  private final Map<String, Receiver<?>> outputDataReceivers;
+  private final Map<KV<String, String>, Receiver<?>> outputTimersReceivers;
+  private final Map<LogicalEndpoint, Long> perEndpointElementCount;
+  private final StreamObserver<Elements> outboundObserver;
+  @VisibleForTesting ScheduledFuture<?> flushFuture;
+  private final LongAdder totalByteCounter = new LongAdder();
+  private final Object flushLock = new Object();
+
+  public BeamFnDataOutboundAggregator(
+      PipelineOptions options,
+      Supplier<String> processBundleRequestIdSupplier,
+      StreamObserver<BeamFnApi.Elements> outboundObserver) {
+    this.sizeLimit = getSizeLimit(options);
+    this.timeLimit = getTimeLimit(options);
+    this.outputDataReceivers = new HashMap<>();
+    this.outputTimersReceivers = new HashMap<>();
+    this.perEndpointElementCount = new HashMap<>();
+    this.outboundObserver = outboundObserver;
+    this.processBundleRequestIdSupplier = processBundleRequestIdSupplier;
+  }
+
+  /** Starts the flushing daemon thread if data_buffer_time_limit_ms is set. */
+  public void startFlushThread() {
+    if (timeLimit > 0 && this.flushFuture == null) {
+      this.flushFuture =
+          Executors.newSingleThreadScheduledExecutor(
+                  new ThreadFactoryBuilder()
+                      .setDaemon(true)
+                      .setNameFormat("DataBufferOutboundFlusher-thread")
+                      .build())
+              .scheduleAtFixedRate(
+                  this::periodicFlush, timeLimit, timeLimit, TimeUnit.MILLISECONDS);
+    }
+  }
+
+  /**
+   * Register the outbound data logical endpoint, returns the FnDataReceiver for processing the
+   * endpoint's outbound data.
+   */
+  public <T> FnDataReceiver<T> registerOutputDataLocation(String pTransformId, Coder<T> coder) {
+    if (outputDataReceivers.containsKey(pTransformId)) {
+      throw new IllegalStateException(
+          "Outbound data endpoint already registered for " + pTransformId);
+    }
+    Receiver<T> receiver = new Receiver<>(coder, totalByteCounter);
+    if (timeLimit > 0) {
+      synchronized (flushLock) {
+        outputDataReceivers.put(pTransformId, receiver);
+        return data -> {
+          checkFlushThreadException();
+          synchronized (flushLock) {
+            receiver.accept(data);
+            if (totalByteCounter.longValue() > sizeLimit) {
+              flush();
+            }
+          }
+        };
+      }
```

Comment:
since the flush thread is only started after registration we don't need this flush lock
```suggestion
        outputDataReceivers.put(pTransformId, receiver);
        return data -> {
          checkFlushThreadException();
          synchronized (flushLock) {
            receiver.accept(data);
            if (totalByteCounter.longValue() > sizeLimit) {
              flush();
            }
          }
        };
```